src/resourceIndex.js

```diff
@@ -0,0 +1,124 @@
+      return;
+    }
+    if (!resources[sourceId][resourceType]) {
+      utils.verbose(`Attempted to access resources for ${sourceId} / ${resourceType} but there are none in the index`);
+      return;
+    }
+    this.data = resources[sourceId][resourceType][resourcePath] || {};
+  }
+
+  /**
+   * Given a YYYY-MM-DD string return one of four values:
+   * - LATEST: if the date is more recent that any in the index
+   * - false: there are no records in the index, or the dateDescriptor
+   *          is mis-formed
+   * - YYYY-MM-DD string: the closest match (may be an exact match)
+   * 
+   * In the future this will be expanded to parse other kinds of
+   * dateDescriptors (e.g. a descriptor meaning "6 months ago").
+   * 
+   * When there is not an exact match we interpret closest to be
+   * the closest but more recent version.
+   */
+  closestDate(dateDescriptor) {
+    if (!dateDescriptor.match(/^\d{4}-\d{2}-\d{2}$/)) {
+      return false;
+    }
+    const availableDates = (this.data?.versions || []).map((v) => v.date);
+    if (!availableDates.length) {
+      return false;
+    }
+    let closest = availableDates[0];
+    if (closest < dateDescriptor) {
+      return LATEST;
+    }
+    for (const date of availableDates) {
+      if (date >= dateDescriptor) {
+        closest = date;
+      } else {
+        break;
+      }
+    }
+    return closest;
+  }
+
+
+  /**
+   * Given a YYYY-MM-DD string, return the available subresources and their corresponding URLs
+   */
+  subresourceUrls(date) {
+    for (const version of this.data?.versions || []) {
+      if (version.date===date) {
+        return version.fileUrls;
+      }
+    }
+    throw new Error(`Attempted to access a specific date which doesn't exist in the resource index: ${date}`)
+  }
+
+}
+
+
+/**
+ * Updates provided data in-place by fetching the file on disk.
+ * We expose an update function rather than automatically updating here. This
+ * prevents scripts / tests which (indirectly) import this file unintentionally
+ * fetching stuff from S3 (and often failing due to missing credentials).
+ */
+async function updateResourceVersions() {
+  let newResources;
+  if (process.env.LOCAL_RESOURCE_INDEX) {
+    utils.verbose(`Updating available resources index from ${process.env.LOCAL_RESOURCE_INDEX}`);
+    let fileContents = fs.readFileSync(process.env.LOCAL_RESOURCE_INDEX)
+    if (process.env.LOCAL_RESOURCE_INDEX.endsWith('.gz')) {
+      fileContents = await gunzip(fileContents)
+    }
+    newResources = JSON.parse(fileContents);
+  } else {
+    const [Bucket, Key] = ['nextstrain-inventories', 'resources.json.gz'];
+    utils.verbose(`Updating available resources index from s3://${Bucket}/${Key}`);
+    const S3 = new AWS.S3();
+    try {
+      const newETag = (await S3.headObject({Bucket, Key}).promise()).ETag;
+      if (newETag === eTag) {
+        utils.verbose("Skipping available resource update as eTag hasn't changed");
+        return;
```

Comment:
I suggest guarding this condition further on `newETag` having a value, e.g. `newETag && newETag === eTag`, in order to avoid an edge case where we fail to get an etag value for whatever reason and so fail to ever update again. That is, if an etag is unavailable, then falling back to updating on every invocation instead of falling back to never updating.

Reply:
No problem, done (although I didn't know it was possible for S3 uploads to not have an eTag).